### PR TITLE
HHH-9965 Pagination ignored on collection fetch join: Add the ability to throw an exception instead of logging a warn

### PR DESCRIPTION
**Pull request by reda-alaoui**
Fix https://hibernate.atlassian.net/browse/HHH-9965

### CODE REVIEW

**Comment by sebersole** [comment]
Awesome! Thanks for doing this

**Comment by reda-alaoui** [comment]
Test case added ;)

File: hibernate-core/src/main/java/org/hibernate/hql/internal/ast/QueryTranslatorImpl.java [hunk 1]
```diff
@@ -377,7 +379,16 @@ public List list(SharedSessionContractImplementor session, QueryParameters query
 
 		QueryParameters queryParametersToUse;
 		if ( hasLimit && containsCollectionFetches() ) {
-			LOG.firstOrMaxResultsSpecifiedWithCollectionFetch();
+			boolean fail = ConfigurationHelper.getBoolean(Environment.FAIL_ON_INTENSIVE_IN_MEMORY_OPERATION,
```
Comment (sebersole) [inline hibernate-core/src/main/java/org/hibernate/hql/internal/ast/QueryTranslatorImpl.java]: You can't do this though.  This ends up being a performance problem.  You end up doing a Map lookup every time a query is translated which the JBoss performance team has consistently identified as a problem - and its unnecessary. 

This really should be integrated with SessionFactoryOptions.  I apologize that doing that is currently pretty painful

Comment (reda-alaoui) [reply hibernate-core/src/main/java/org/hibernate/hql/internal/ast/QueryTranslatorImpl.java]: Done !

File: hibernate-core/src/main/java/org/hibernate/boot/internal/SessionFactoryBuilderImpl.java [hunk 2]
```diff
@@ -66,61 +66,7 @@
...
-import static org.hibernate.cfg.AvailableSettings.USE_MINIMAL_PUTS;
-import static org.hibernate.cfg.AvailableSettings.USE_QUERY_CACHE;
-import static org.hibernate.cfg.AvailableSettings.USE_SCROLLABLE_RESULTSET;
-import static org.hibernate.cfg.AvailableSettings.USE_SECOND_LEVEL_CACHE;
-import static org.hibernate.cfg.AvailableSettings.USE_SQL_COMMENTS;
-import static org.hibernate.cfg.AvailableSettings.USE_STRUCTURED_CACHE;
-import static org.hibernate.cfg.AvailableSettings.VALIDATE_QUERY_PARAMETERS;
-import static org.hibernate.cfg.AvailableSettings.WRAP_RESULT_SETS;
+import static org.hibernate.cfg.AvailableSettings.*;
 import static org.hibernate.engine.config.spi.StandardConverters.BOOLEAN;
```
Comment (sebersole) [inline hibernate-core/src/main/java/org/hibernate/boot/internal/SessionFactoryBuilderImpl.java]: Please undo this part.  Try to change just relevant lines.  See [here](https://github.com/hibernate/hibernate-ide-codestyles) for codestyle definitions for you IDE of choice

File: hibernate-core/src/main/java/org/hibernate/cfg/AvailableSettings.java [hunk 3]
```diff
@@ -1810,4 +1810,10 @@
 	 *
 	 */
 	String TABLE_GENERATOR_STORE_LAST_USED = "hibernate.id.generator.stored_last_used";
+
+	/**
+	 * Raises an exception when an in-memory intensive operation is about to be performed.
+	 * Disabled by default. Set to true to enable.
+	 */
+	String FAIL_ON_INTENSIVE_IN_MEMORY_OPERATION = "hibernate.fail_on_intensive_in_memory_operation";
```
Comment (sebersole) [inline hibernate-core/src/main/java/org/hibernate/cfg/AvailableSettings.java]: IMO this name is way too vague.  I'd prefer to name this something more concrete, like `hibernate.query.fail_on_pagination_over_collection_fetch`

File: hibernate-core/src/main/java/org/hibernate/boot/internal/SessionFactoryBuilderImpl.java (hunk 2 above)
Comment (reda-alaoui) [reply hibernate-core/src/main/java/org/hibernate/boot/internal/SessionFactoryBuilderImpl.java]: Done

File: hibernate-core/src/main/java/org/hibernate/cfg/AvailableSettings.java (hunk 3 above)
Comment (reda-alaoui) [reply hibernate-core/src/main/java/org/hibernate/cfg/AvailableSettings.java]: Done

**Comment by reda-alaoui** [comment]
Thank you !

**Pull request merged**